src/rgw/rgw_op.cc

```diff
@@ -2263,12 +2263,23 @@ void RGWGetObj::execute(optional_yield y)
     try {
       decode(m, attr_iter->second);
       if (m.get_tier_type() == "cloud-s3") {
-        /* XXX: Instead send presigned redirect or read-through */
-        op_ret = -ERR_INVALID_OBJECT_STATE;
-        s->err.message = "This object was transitioned to cloud-s3";
-        ldpp_dout(this, 4) << "Cannot get cloud tiered object. Failing with "
-            << op_ret << dendl;
-        goto done_err;
+        if (!sync_cloudtiered) {
+          /* XXX: Instead send presigned redirect or read-through */
+          op_ret = -ERR_INVALID_OBJECT_STATE;
+          s->err.message = "This object was transitioned to cloud-s3";
+          ldpp_dout(this, 4) << "Cannot get cloud tiered object. Failing with "
+              << op_ret << dendl;
+          goto done_err;
+        } else { // fetch object for sync and set cloud_tier attrs
+          bufferlist t, t_tier;
+          RGWObjTier tier_config;
+          m.get_tier_config(&tier_config);
+
+          encode("cloud-s3", t);
```

Comment:
you can just call `t.append("cloud-s3")`. `encode()` would add an unnecessary length prefix

on the other end, you can use `bufferlist::to_str()` instead of `decode()`